### PR TITLE
add sr_folder_ref_dataset

### PR DESCRIPTION
**Pull request by Yshuo-Li**
For TTSR

### CODE REVIEW

File: mmedit/datasets/sr_folder_ref_dataset.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        │   ├── 0002_x4.png
+
+    then, you need to set:
+
+    .. code-block:: python
+
+        ref_folder = 'data_root/ref'
+        gt_folder = 'data_root/gt'
+        lq_folder = 'data_root/lq'
+        filename_tmpl = '{}_x4'
```
Comment (ckkelvinchan) [inline mmedit/datasets/sr_folder_ref_dataset.py:46]:
```
filename_tmpl_gt = '{}'
filename_tmpl_lq = '{}_x4'
```

File: mmedit/datasets/sr_folder_ref_dataset.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        self.gt_folder = str(gt_folder) if gt_folder else None
+        self.lq_folder = str(lq_folder) if lq_folder else None
+        self.filename_tmp_gt = filename_tmp_gt
+        self.filename_tmp_lq = filename_tmp_lq
+        self.data_infos = self.load_annotations()
+
+    def load_annotations(self):
+        """Load annoations for SR dataset.
+
+        It loads the ref and GT image path from folders.
```
Comment (ckkelvinchan) [inline mmedit/datasets/sr_folder_ref_dataset.py:85]: It loads the ref, GT, and LQ image path from folders.

File: mmedit/datasets/sr_folder_ref_dataset.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+        Returns:
+            dict: Returned dict for ref and GT pairs.
```
Comment (ckkelvinchan) [inline mmedit/datasets/sr_folder_ref_dataset.py:88]: ref, GT, and LQ triplets

File: mmedit/datasets/sr_folder_ref_dataset.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+
+@DATASETS.register_module()
+class SRFolderRefDataset(BaseSRDataset):
+    """
+    General paired image folder dataset for reference-based image restoration.
+
+    The dataset loads ref (reference) image pairs
+        Must contain: ref (reference)
+        Optional: GT (Ground-Truth), LQ (Low Quality), or both
```
Comment (ckkelvinchan) [inline mmedit/datasets/sr_folder_ref_dataset.py:14]: Is there any case where only reference is needed? I think at least GT should be required?

File: tests/test_datasets.py [hunk 5]
```diff
@@ -280,6 +281,60 @@ def test_sr_folder_gt_dataset(self):
...
+        assert data_infos == [
+            dict(
+                lq_path=str(lq_folder / 'baboon_x4.png'),
+                gt_path=str(gt_folder / 'baboon.png'),
+                ref_path=str(ref_folder / 'baboon.png'))
+        ]
+        result = sr_folder_ref_dataset[0]
+        assert len(sr_folder_ref_dataset) == 1
+        assert check_keys_contain(result.keys(), target_keys)
+
```
Comment (ckkelvinchan) [inline tests/test_datasets.py:337]: unittests of the assertion errors?

File: mmedit/datasets/sr_folder_ref_dataset.py (hunk 4 above)
Comment (Yshuo-Li) [reply mmedit/datasets/sr_folder_ref_dataset.py:14]: One of the follows:
- Ref and LQ
- Ref and GT,
- Ref, LQ, and GT,

Comment (ckkelvinchan) [reply mmedit/datasets/sr_folder_ref_dataset.py:14]: Would it be better to raise error if `lq_folder` and `gt_folder` are both `None`?

File: mmedit/datasets/sr_folder_ref_dataset.py [hunk 6]
```diff
@@ -72,6 +73,7 @@ def __init__(self,
                  filename_tmp_gt='{}',
                  filename_tmp_lq='{}'):
         super().__init__(pipeline, scale, test_mode)
+        assert gt_folder or lq_folder
```
Comment (ckkelvinchan) [inline mmedit/datasets/sr_folder_ref_dataset.py:76]: Better add a message here (e.g. `At least one of 'gt_folder and 'lq_folder' cannot be None.`). Looks good to me other than this.

**Review by ckkelvinchan — APPROVED** [review]

File: mmedit/datasets/sr_folder_ref_dataset.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                 filename_tmp_gt='{}',
+                 filename_tmp_lq='{}'):
+        super().__init__(pipeline, scale, test_mode)
+        assert gt_folder or lq_folder, 'At least one of gt_folder and' \
+            'lq_folder cannot be None.'
+        self.scale = scale
+        self.ref_folder = str(ref_folder)
+        self.gt_folder = str(gt_folder) if gt_folder else None
+        self.lq_folder = str(lq_folder) if lq_folder else None
+        self.filename_tmp_gt = filename_tmp_gt
```
Comment (innerlee) [inline mmedit/datasets/sr_folder_ref_dataset.py:82]: Abbreviation for template is `tmpl`
tmp often means temporarily

**Review by innerlee — APPROVED** [review]

**Pull request merged**